Tests/SonarScanner.MSBuild.PreProcessor.Test/WebClientDownloaderTest.cs

```diff
@@ -146,26 +150,156 @@ public async Task Download_Fail_NotForbidden_WithLog()
             var logger = new TestLogger();
             var sut = CreateSut(logger, HttpStatusCode.NotFound);
 
-            var text = await sut.Download(TestUri, true);
+            var text = await sut.Download(RelativeUrl, true);
 
             text.Should().BeNull();
-            logger.AssertDebugLogged("Downloading from https://www.sonarsource.com/...");
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeOk_SucceedWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.OK);
+
+            var result = await sut.TryDownloadIfExists(RelativeUrl);
+
+            result.Item1.Should().BeTrue();
+            result.Item2.Should().Be(TestContent);
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeNotFound_SucceedWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.NotFound);
+
+            var result = await sut.TryDownloadIfExists(RelativeUrl);
+
+            result.Item1.Should().BeFalse();
+            result.Item2.Should().BeNull();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeForbidden_ShouldThrowWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+
+            Func<Task> act = async () => await sut.TryDownloadIfExists(RelativeUrl);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertNoWarningsLogged();
+        }
+
+        [TestMethod]
+        public async Task TryDownloadIfExists_HttpCodeForbiddenWithWarnLog_ShouldThrowWithLog()
+        {
+            var logger = new TestLogger();
+            var sut = CreateSut(logger, HttpStatusCode.Forbidden);
+
+            Func<Task> act = async () => await sut.TryDownloadIfExists(RelativeUrl, true);
+
+            await act.Should().ThrowExactlyAsync<HttpRequestException>();
+            logger.AssertDebugLogged($"Downloading from {BaseUrl}{RelativeUrl}...");
+            logger.AssertWarningLogged("To analyze private projects make sure the scanner user has 'Browse' permission.");
+        }
+
+        public static IEnumerable<object[]> UnsuccessfulHttpCodeData =>
```

Comment:
this should be under the tests but I would remove it and test with a single value. I'm not sure if checking with all of them brings value. Maybe I'm missing something.